pyout/common.py

```diff
@@ -0,0 +1,395 @@
+        # We need a second pass with strip_callables because norm_row will
+        # contain new callables for any delayed values.
+        callables1 = self.strip_callables(norm_row)
+        return callables0 + callables1, norm_row
+
+    def _maybe_delay(self, getter, row, columns):
+        row_norm = {}
+        for column in columns:
+            if column not in self.delayed_columns:
+                row_norm[column] = getter(row, column)
+
+        def delay(cols):
+            return lambda: {c: getter(row, c) for c in cols}
+
+        for columns in self.delayed.values():
+            key = columns[0] if len(columns) == 1 else tuple(columns)
+            row_norm[key] = delay(columns)
+        return row_norm
+
+    @staticmethod
+    def strip_callables(row):
+        """Extract callable values from `row`.
+
+        Replace the callable values with the initial value (if specified) or
+        an empty string.
+
+        Parameters
+        ----------
+        row : mapping
+            A data row.  The keys are either a single column name or a tuple of
+            column names.  The values take one of three forms: 1) a
+            non-callable value, 2) a tuple (initial_value, callable), 3) or a
+            single callable (in which case the initial value is set to an empty
+            string).
+
+        Returns
+        -------
+        list of (column, callable)
+        """
+        callables = []
+        to_delete = []
+        to_add = []
+        for columns, value in row.items():
+            if isinstance(value, tuple):
+                initial, fn = value
+            else:
+                initial = NOTHING
+                # Value could be a normal (non-callable) value or a
+                # callable with no initial value.
+                fn = value
+
+            if callable(fn) or inspect.isgenerator(fn):
+                if not isinstance(columns, tuple):
+                    columns = columns,
+                else:
+                    to_delete.append(columns)
+                for column in columns:
+                    to_add.append((column, initial))
+                callables.append((columns, fn))
+
+        for column, value in to_add:
+            row[column] = value
+        for multi_columns in to_delete:
+            del row[multi_columns]
+
+        return callables
+
+    # Input-specific getters.  These exist as their own methods so that they
+    # can be wrapped in a callable and delayed.
+
+    def getter_dict(self, row, column):
+        return row.get(column, self.nothings[column])
+
+    def getter_seq(self, row, column):
+        col_to_idx = {c: idx for idx, c in enumerate(self._columns)}
+        return row[col_to_idx[column]]
+
+    def getter_attrs(self, row, column):
+        return getattr(row, column, self.nothings[column])
+
+
+def _safe_get(mapping, key, default=None):
+    try:
+        return mapping.get(key, default)
+    except AttributeError:
+        return default
```

Comment:
could you please add a docstring describing the usecases where this would be actually needed?   to me it just sounds like a dangerous "if I assume it is a mapping but it is not a mapping - then return `default`" -- shouldn't that be decided ahead either at that point you are dealing with a mapping or not? just not sure why

Reply:
> could you please add a docstring describing the usecases where this would be actually needed?

Sure, will add.  It's an internal helper only used in the context of styles.  It avoids the common "is this a mapping" check that happens because 1) some style values can be a single atom (like "width": "auto") or a mapping ("width": {"auto": True, "min"}...).

> to me it just sounds like a dangerous "if I assume it is a mapping but it is not a mapping - then return default" -- shouldn't that be decided ahead either at that point you are dealing with a mapping or not? just not sure why

I agree in general.  I'm less worried here because the styles are validated.  Let me know if you have other suggestions though.